pkg/log/progress/doc.go

```
@@ -0,0 +1,40 @@
+// Copyright (c) 2022, Oracle and/or its affiliates.
+// Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl.
+
+package progress
+
+// The progress package provides logging for progress messages in controllers and other
+// code that wants to throttle messages.  The Verrazzano Platform Operator controller for
+// the Verrazzano resource is an example where this should be used.  During a reconcile cycle,
+// the controller-runtime calls the controller Reconcile method repeatedly until the resource has
+// been reconciled.  For something like Verrazzano install, this can translate into scores of
+// reconcile calls.  The controller Reconcile code has no way of knowing if it already displayed
+// an info message, such as "Keyclaok waiting for Verrazzano secret".  Using normal logging, that message
+// might be displayed dozens of times.  With the progress logger, the message is only displayed
+// periodically, once a minute by default.  Furthermore, once a message is display, and a new message
+// is logged, that old message will never be displayed again.  This allows controllers to log
+// informative progress messages, without overwhelming the log files with superfluous information.
+//
+// This logger is initialized with the zap.SugaredLogger and then used instead of the zap logger directly.
+// The same SugaredLogger calls can be made: Debug, Debugf, Info, Infof, Error, and Errorf.  The
+// two new calls are Progress and Progressf. The S() method will return the underlying SugaredLogger.
+// The following psuedo-code shows how this should be used:
+//
+//   l := EnsureLogger("namespace/myresource", zap.S()).DefaultProgressLogger()
+//
+// Display info and errors as usual
+//   p.Errorf(...)
```

What's the justification behind supporting error messages in this format? I'd assume we want progress messages for Reconciliation Info. It seems odd to ignore error messages.

---

IIUC it will throttle repeated messages

* IFF the message is logged using `Progress` 
* the same exact message is being recorded within the throttling interval

Is that correct?  If so, it's up to the caller to decide what message to throttle and which ones not to, so the caller would have to be careful not to throttle things that shouldn't be.

---

yes that is correct.  I think most of the stuff our controllers log would be progress messages, since reconcile gets called repeatedly and we have no state.

---

Also, we don't ignore error message, the info, error, and debug messages are still logged directly by zap

---

I thought about changing the info sematics to be progress but thought it was better for the code to explicitly log progress messages.